docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Multiple attributes exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+      source: https://github.com/knative/eventing/pulls/123
+```
+
+Specified with a CEL expression:
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Prefix match on source
+
+Cannot be specified with the SourceAndType field.
+
+CEL doesn't support prefix match directly, but prefix match can be accomplished
+with a regular expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.source.matches("^https://github.com")
+```
+
+### Complex boolean expression
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" ||
+        (ce.type == "com.github.issue.create" && ce.source.matches("proposals")
+
+```
+
+### Parsed extensions
+
+Cannot be specified with the SourceAndType field.
+
+_This example assumes the repository name is available as a CloudEvents
+extension with name `repository`._
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseExtensions: true
+      expression: >
+        ce.type == "com.github.pull.create" ||
+         (ce.type == "com.github.issue.create" &&
+          ce.repository == "proposals")
+
+```
+
+### Parsed data
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+_Until [google/cel-go#203](https://github.com/google/cel-go/issues/203) is
+resolved, dynamic integer fields in CEL must be compared as floats._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseData: true
+      expression: >
+        ce.type == "dev.knative.observation" &&
+        data.latency > 300.0
+```
+
+## Caveats
+
+### CEL
+
+The main caveat is the choice of CEL as expression language.
+
+CEL is new and the language is not officially supported by Google. Participation
+by non-Google contributors is low. OPA is planning to replace their policy
+language Rego with CEL, but
+[no timeline has been announced for this migration](https://github.com/open-policy-agent/opa/issues/811#issuecomment-401844999).
+The KrakenD API Gateway has announced support for CEL
+[here](https://medium.com/devops-faith/krakend-api-gateway-0-9-released-9427c249dbcd).
+
+The CEL language is new and has limited exposure, despite being informed by
+experience and research at Google. We may discover issues that make it less
+suitable for Trigger filtering (e.g.
+[google/cel-go#203](https://github.com/google/cel-go/issues/203).
```

Comment:
In the ideal case the user is rarely using CEL expressions directly, but your underlying evaluation model can leverage CEL to support a variety of data-driven formats.

The following filter can be evaluated under the covers by a well-defined CEL expression:
```yaml
filter:
   matchTypes: "com.github.pull.create"
   matchSource: "https://github.com/*"
```

Filter enforcement in CEL:
```go
(!has(filter.matchTypes) || match(ce.type, filter.matchTypes)) &&
(!has(filter.matchSource) || match(ce.source, filter.matchSource))
```

Here, CEL is just used under the covers, which would allow you to support a variety of different filter formats with very little implementation overhead. 

Note, probably a good idea to version the filter or specify the filter type in order to version evaluation semantics cleanly over time.